강재현 1주차

Recreate-Kang opened the pull request:
강재현-1주차

Recreate-Kang committed:
feat : sandbox CRUD <Get, Post, Patch, Delete> add

Recreate-Kang committed:
refactor : 공통 매핑/todos path를 requestmapping으로 통합

Recreate-Kang committed:
refactor : 주석 삭제 및 추가, {파일클래스명, 폴더명} 변경

jisu-0305 commented on sandbox/src/main/resources/application.properties:
```diff
@@ -0,0 +1,5 @@
+spring.application.name=sandbox
+spring.datasource.url=jdbc:mysql://localhost:3306/sandbox
+spring.datasource.username=root
+spring.datasource.password=root
```
당장 사용하는 DB가 local DB라서 큰 상관은 없지만 추후 DB 서버를 이용하실 경우에는 git에 올라오지 않게 아래 내용을 참고해서 별도로 관리하시면 좋을 것 같습니다.
https://itinerant.tistory.com/202

Recreate-Kang replied:
네 감사합니다.
중요한 키는 따로 관리하는게 옳은거 같네요.
주신 자료 잘 참고하도록하겠습니다

jisu-0305 commented on sandbox/src/main/java/com/ssafy/sandbox/respository/TodoRepository.java:
```diff
@@ -0,0 +1,22 @@
+
+public interface TodoRepository extends JpaRepository<TodoDto, Integer> {
+
+    List<TodoDto> findAll();
```
일전에 TodoDto에 테이블 어노테이션과 엔티티 어노테이션을 통해 연결을 지어줬고 그에 따라 findAll은 JPARepository기본 제공 메서드로 알고 있어 직접 안적어도 되는걸로 알고 있습니다.
하단의 블로그 참고하시면 좋을 것 같습니다.
https://velog.io/@knavoid/Spring-Data-JPA

Recreate-Kang replied:
자료 감사합니다.
참고해서 수정하도록하겠습니다.

jisu-0305 commented on sandbox/src/main/java/com/ssafy/sandbox/respository/TodoRepository.java:
```diff
@@ -0,0 +1,22 @@
+    void toggleTodoCompletion(@Param("todoId") int todoId);
+    @Modifying
+    @Override
+    void deleteById(Integer todoID);
```
위와 동일한 이유로 생략 가능합니다.

jisu-0305 commented on sandbox/src/main/java/com/ssafy/sandbox/dto/TodoDto.java:
```diff
@@ -0,0 +1,26 @@
+    @Column(name = "user_id")
+    int userId;
+
+    public TodoDto() {
+
+    }
```
data 어노테이션이 생성자 자동생성해주는 RequiredArgsConstructor의 기능도 해주는 것으로 알고 있어서 생성자 생략 가능합니다.

jisu-0305 reviewed (commented):
고생하셨습니다.

T0nixx commented on sandbox/src/main/java/com/ssafy/sandbox/config/WebConfig.java:
```diff
@@ -10,7 +10,8 @@ public class WebConfig implements WebMvcConfigurer {
                 .allowedOrigins("https://ssafysandbox.vercel.app")
-                .allowedMethods("GET", "POST", "PUT", "DELETE", "OPTIONS")
-                .allowedHeaders("*");
+                .allowedMethods("GET", "POST", "PUT", "DELETE", "OPTIONS","PATCH")
```
Code formatting에 대해 공부해봅시다 :)

T0nixx commented on sandbox/src/main/java/com/ssafy/sandbox/controller/TodoController.java:
```diff
@@ -1,44 +1,50 @@
+    public ResponseEntity<TodoDto> createTodo(@RequestBody TodoDto newTodoDto) {
+        newTodoDto.setCompleted(false); // 기본값 설정
+        newTodoDto.setCreatedAt(LocalDateTime.now().toString()); // 현재 시간 설정
+        newTodoDto.setUserId(1); // 사용자 ID 설정 (예시)
```
해당 부분의 setter들은 TodoDto를 너무 재사용하려고 한 것 같아요.
endpoint 별로 필요한 Dto가 다른 편이 더 쉽게 이해할 수 있을 것 같은데 어떻게 생각하세요?

Recreate-Kang replied:
하나의 DTO를 활용해서 재사용성을 높이는게 좋다는 생각을 했는데.
용도에 맞는 DTO를 생성해서 사용하는것이 타당한거 같습니다.

T0nixx commented on sandbox/src/main/java/com/ssafy/sandbox/controller/TodoController.java:
```diff
@@ -1,44 +1,50 @@
+        todoService.toggleTodo(todoId); // 서비스 메서드 호출
+        Map<String, Boolean> update = new HashMap<>();
+        update.put("complete", Boolean.TRUE);
+        return ResponseEntity.ok().body(update);
```
getTodos()에서는 ResponseEntity.ok() 안에 response를 넣었는데 여기선 body를 한번 더 호출한 이유가 있나요?

T0nixx commented on sandbox/src/main/java/com/ssafy/sandbox/controller/TodoController.java:
```diff
@@ -1,44 +1,50 @@
+    @DeleteMapping("/{todoID}")
+    public ResponseEntity<Void> deleteTodo( @PathVariable int todoID) {
+        todoService.deleteTodo(todoID);
+        return ResponseEntity.ok().build();
```
response 가 비어있는 경우에는 ok 보다 noContent로 응답을 주는 편이 http 코드 정의에 더 부합할 것 같아요

T0nixx commented on sandbox/src/main/java/com/ssafy/sandbox/dto/TodoDto.java:
```diff
@@ -0,0 +1,26 @@
+    String content;
+    @Column
+    boolean completed;
+    @Column(name = "created_at", columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP")
```
시간을 String으로 저장하는 것 보다는 LocalDateTime이나 Instant, ZonedDateTime을 통해 관리하면 좋을 것 같아요.

Recreate-Kang committed:
refactor: 1. 컨트롤러(createdTodo__메서드의 setter를 Service로 옮김
          2. 컨트롤러(공통)_ResponseEntity의 build를 status와 body를 분리함
          3. 리포지토리_JPA 지원 extends 메서드 중복 삭제
          4. TodoDto_의 created_at 타입을 String->LocalDateTime으로 변경

Recreate-Kang committed:
refactor: 1. 컨트롤러(createdTodo__메서드의 setter를 Service로 옮김
          2. 컨트롤러(공통)_ResponseEntity의 build를 status와 body를 분리함
          3. 리포지토리_JPA 지원 extends 메서드 중복 삭제
          4. TodoDto_의 created_at 타입을 String->LocalDateTime으로 변경

Recreate-Kang committed:
refactor: 1. 컨트롤러(toggleTodo)_메서드의 Msg생성을 Service로 옮김
          2. 컨트롤러(getTodos)_메서드의 Key-Value 방식을 ResponseDto로 형식 변경
feat : 1. DTO(TodoListResponse)_생성, <String, List<TodoDto>>를 대체

Recreate-Kang committed:
refactor: 1. 데이터베이스 노출된 값들 비공개 및 별도관리

Recreate-Kang committed:
refactor: TodoService(createTodos)의 sql에서 처리되어 불필요한 연산 제거

Recreate-Kang committed:
refactor: TodoService(createTodos)의 sql에서 처리되어 불필요한 연산 제거
